src/polynomial.cpp

```diff
@@ -0,0 +1,307 @@
+#include "polynomial.h"
+#include "add.h"
+#include "mul.h"
+#include "pow.h"
+#include "constants.h"
+
+namespace SymEngine {
+
+Polynomial::Polynomial(const RCP<const Symbol> &var, map_uint_mpz&& dict) :
+     var_{var}, dict_{std::move(dict)} {
+
+    map_uint_mpz::iterator it = dict_.begin();
+    uint largest = it->first;
+    uint cur;
+    while (it != dict_.end()) {
+        cur =it->first;
+        if (cur > largest)
+            largest = cur;
+        it++;
+    }
+    degree = largest;
+
+    SYMENGINE_ASSERT(is_canonical(degree, dict_))
+}
+
+Polynomial::Polynomial(const RCP<const Symbol> &var, const std::vector<mpz_class> &v) :
+     var_{var} {
+
+    for (uint i = 0; i < v.size(); i++) {
+        if (v[i] != 0)
+            dict_add_term(dict_, v[i], i);
+    }
+    degree = v.size() - 1;
+
+    SYMENGINE_ASSERT(is_canonical(degree, dict_))
+}
+
+bool Polynomial::is_canonical(const uint &degree, const map_uint_mpz& dict)
+{
+    map_uint_mpz ordered(dict.begin(), dict.end());
+    uint prev_degree = (--ordered.end())->first;
+    if (prev_degree != degree)
+        return false;
+
+    return true;
+}
+
+std::size_t Polynomial::__hash__() const 
+{
+    std::hash<std::string> hash_string;
+    std::hash<uint> hash_uint;
+    std::hash<long long int> hash_si;
+    std::size_t seed = POLYNOMIAL;
+
+    seed += hash_string(this->var_->get_name());
+    for (auto &it : this->dict_)
+    {
+        seed += hash_uint(it.first) + hash_si(it.second.get_si());
```

Comment:
You should combine the hashes for `it.first` and `it.second`, not just with a `+`. The only place you use `+` is for combining hashes for different `it`, as you do. Use the `hash_combine` function.